File: C++/섹션07. BFS - 2차원 배열 탐색/01. BOJ 2667 - 단지번호붙이기.cpp
#include <stdio.h>

#define MAX (25 + 5)

int N;

int MAP[MAX][MAX];
bool visit[MAX][MAX];

int answers[MAX * MAX];

struct RC
{
	int r;
	int c;
};

RC queue[MAX * MAX];

// ↑, →, ↓, ←
int dr[] = { -1, 0, 1, 0 };
int dc[] = { 0, 1, 0, -1 };

void printVisit() // for debug
{
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
			printf("%d ", visit[r][c]);
		putchar('\n');
	}
	putchar('\n');
}

void input()
{
	scanf("%d", &N);

	for (int r = 1; r <= N; r++)
		for (int c = 1; c <= N; c++)
			scanf("%1d", &MAP[r][c]);
}

int BFS(int r, int c)
{
	// printf("start %d, %d\n", r, c);

	int rp, wp;

	rp = wp = 0;

	queue[wp].r = r;
	queue[wp++].c = c;
	
	visit[r][c] = true;
	// printVisit();

	while (rp < wp)
	{
		RC out = queue[rp++];

		// printf("out %d, %d\n", out.r, out.c);

		for (int i = 0; i < 4; i++)
		{
			int nr, nc;

			nr = out.r + dr[i];
			nc = out.c + dc[i];

			if (MAP[nr][nc] == 0 || visit[nr][nc] == true) continue;

			queue[wp].r = nr;
			queue[wp++].c = nc;

			visit[nr][nc] = true;

			// printf("push %d, %d\n", nr, nc);
			// printVisit();
		}
	}

	return wp;
}

int main()
{
	input();

	int ansCount = 0;
	for (int r = 1; r <= N; r++)
	{
		for (int c = 1; c <= N; c++)
		{
			if(MAP[r][c] == 1 && visit[r][c] == false)
				answers[ansCount++] = BFS(r, c);
		}
	}

	printf("%d\n", ansCount);

	for (int i = 0; i < ansCount - 1; i++)
	{
		for (int k = i + 1; k < ansCount; k++)
		{
			if (answers[i] > answers[k])
			{
				int tmp = answers[i];
				answers[i] = answers[k];
				answers[k] = tmp;
			}
		}
	}

	for (int i = 0; i < ansCount; i++) printf("%d\n", answers[i]);

	return 0;
}

File: Python/섹션07. BFS - 2차원 배열 탐색/01. BOJ 2667 - 단지번호붙이기.py
MAX = 25 + 5

N = int(input())

MAP = [[0] * MAX for _ in range(MAX)]
visit = [[False] * MAX for _ in range(MAX)]

class RC:
    def __init__(self, r: int, c: int):
        self.r = r
        self.c = c

queue = [RC(0, 0) for _ in range(MAX * MAX)]   

# ↑, →, ↓, ←
dr = [-1, 0, 1, 0]
dc = [0, 1, 0, -1]

def print_visit():
    for r in range(1, N + 1):
        print(*list(map(int, visit[r][1:N + 1])))

def input_data():
    for r in range(1, N + 1):
        MAP[r][1:N + 1] = list(map(int, input()))
        
def BFS(r, c):
    rp, wp = 0, 0
    
    queue[wp].r = r
    queue[wp].c = c
    wp += 1
    
    visit[r][c] = True
        
    while rp < wp:
        out = queue[rp]
        rp += 1
        
        for i in range(4):
            nr = out.r + dr[i]
            nc = out.c + dc[i]
            
            if MAP[nr][nc] == 0 or visit[nr][nc] == True: continue
            
            queue[wp].r = nr
            queue[wp].c = nc
            wp += 1
            
            visit[nr][nc] = True
                
    return wp

input_data()

answer = []
for r in range(1, N + 1):
    for c in range(1, N + 1):
        if MAP[r][c] == 1 and visit[r][c] == False:
            answer.append(BFS(r, c))

print(len(answer))

answer = sorted(answer)

for ans in answer:
    print(ans)